src/CoreWCF.Primitives/src/CoreWCF/IdentityModel/Selectors/WindowsSecurityTokenAuthenticator.cs

```diff
@@ -31,6 +38,14 @@ protected override ReadOnlyCollection<IAuthorizationPolicy> ValidateTokenCore(Se
         {
             var windowsToken = (WindowsSecurityToken)token;
             var claimSet = new WindowsClaimSet(windowsToken.WindowsIdentity, windowsToken.AuthenticationType, _includeWindowsGroups, windowsToken.ValidTo);
+            if (_ldapSettings != null && claimSet.Count > 0)
+            {
```

Comment:
On .NET Framework, it's WindowsClaimSet which handles retrieving all the claims. I think it would be worth adding an overload to the constructory to pass the _ldapSettings instead and have that do the work. That way the code to call RetrieveClaims doesn't need to be duplicated.

Reply:
I went through the code with approach discussed but found the problem is WindowsClaimSet is very much tied with WindowsIdentity, and we can't initialize WindowsIdentity with SPN, we can do by UPN only. I can make changes inside WindowsClaimSet to act differently when it's WIndows/Linux, but that would lead to lot of if/else code for various OS and to me it won't be nice... so I would suggest to have GenericIdentity as we had already decided and in the similar line in .net core I believe.. if you think otherwise, we can discuss and close on this... keeping open to discuss..